File: packages/node/src/api/Logger.ts
import { Timer, TimerContext } from './Timer';
import { LogContext, LogEvent } from './types';

export interface Logger {
  debug(message: string, context?: Record<string, unknown>): void;
  info(message: string, context?: Record<string, unknown>): void;
  warn(message: string, context?: Record<string, unknown>): void;
  error(message: string, error?: Error, context?: Record<string, unknown>): void;

  // PUBLIC_INTERFACE
  /**
   * Export a FATAL log event through the normal processing pipeline.
   *
   * @param message - Human-readable event message.
   * @param context - Optional structured event context.
   */
  fatal(message: string, context?: Record<string, unknown>): void;

  // PUBLIC_INTERFACE
  /**
   * Process a canonical log event through enrichment, masking, validation,
   * batching, and the configured exporter.
   *
   * @param event - Canonical event fields. The object is never mutated.
   * @throws TypeError when the event or its message is invalid.
   * @throws Error when the severity is unsupported or the logger is closed.
   */
  event(event: LogEvent): void;

  // PUBLIC_INTERFACE
  /**
   * Create an unstarted timer that emits an INFO completion event through this logger.
   *
   * @param name - Non-blank canonical event name for the completion event.
   * @param context - Optional event metadata merged with active scoped context on completion.
   * @returns A new unstarted timer.
   * @throws Error when this logger is closed or the name is invalid.
   */
  createTimer(name: string, context?: TimerContext): Timer;

  // PUBLIC_INTERFACE
  /**
   * Run a synchronous or asynchronous callback with execution-scoped log context.
   *
   * Context propagates through promises and async/await. Nested calls inherit
   * unspecified values and restore the preceding scope after completion.
   *
   * @param context - Trace, correlation, and tenant identifiers for the execution scope.
   * @param callback - Work to execute while the supplied context is active.
   * @returns The callback result.
   */
  withContext<T>(context: LogContext, callback: () => T): T;

  /**
   * Wait for delivery work that was in progress when flushing began.
   *
   * @returns A promise that resolves after pending exporter work settles, or
   * rejects when delivery errors are configured to fail closed.
   */
  flush(): Promise<void>;

  /**
   * Close the logger and prevent future logging or flush operations.
   *
   * This operation is idempotent.
   */
  close(): void;
}

File: packages/python/dt3_api/logger.py
"""Public logging contract for the DT3 Python SDK."""

from __future__ import annotations

from typing import Any, Dict, Mapping, Optional, Protocol

from .timer import Timer


class Logger(Protocol):
    """Public interface for emitting DT3 log events and creating timers."""

    # PUBLIC_INTERFACE
    def debug(self, message: str, context: Optional[Dict[str, Any]] = None) -> None:
        """Emit a DEBUG event."""

    # PUBLIC_INTERFACE
    def info(self, message: str, context: Optional[Dict[str, Any]] = None) -> None:
        """Emit an INFO event."""

    # PUBLIC_INTERFACE
    def warn(self, message: str, context: Optional[Dict[str, Any]] = None) -> None:
        """Emit a WARN event."""

    # PUBLIC_INTERFACE
    def error(
        self,
        message: str,
        error: Optional[Exception] = None,
        context: Optional[Dict[str, Any]] = None,
    ) -> None:
        """Emit an ERROR event with optional structured exception details."""

    # PUBLIC_INTERFACE
    def fatal(self, message: str, context: Optional[Dict[str, Any]] = None) -> None:
        """Emit a FATAL event."""

    # PUBLIC_INTERFACE
    def event(self, event_object: Mapping[str, Any]) -> None:
        """Process a canonical event through the normal logger pipeline."""

    # PUBLIC_INTERFACE
    def flush(self) -> None:
        """Flush pending log events to configured exporters."""

    # PUBLIC_INTERFACE
    def close(self) -> None:
        """Close the logger and its configured exporter."""

    # PUBLIC_INTERFACE
    def register_sink(self, sink: Any, name: Optional[str] = None) -> str:
        """Register an additional event sink for runtime fan-out."""

    # PUBLIC_INTERFACE
    def create_timer(
        self,
        name: str,
        context: Optional[Dict[str, Any]] = None,
    ) -> "Timer":
        """Create an unstarted Timer that emits a completion event when stopped."""
